Tutorial: Update Next.js app route (#12178)
PostHog/posthog.com: contents/tutorials/nextjs-app-directory-analytics.md

---
title: 'How to set up Next.js app router analytics, feature flags, and more'
date: 2025-07-15
author:
  - ian-vanagas
showTitle: true
sidebar: Docs
tags:
  - configuration
  - feature flags
  - events
---

Next.js is one of the most popular frameworks for building web apps. Have one and need to know what users are doing in these apps or release a feature safely? PostHog can help.

In this tutorial, we'll create a simple Next.js app and set up PostHog on the client and server side. We'll also capture pageviews and custom events, set up feature flags, and more.

> <Callout type="info" title="Looking for the pages router?">
>   If you use Next.js with the **pages** router, check out our other [Next.js pages router analytics tutorial](/tutorials/nextjs-analytics).
> </Callout>

Author: ```suggestion
import { CalloutBox } from 'components/Docs/CalloutBox'

<CalloutBox type="info" title="Looking for the pages router?">

    If you use Next.js with the **pages** router, check out our other [Next.js pages router analytics tutorial](/tutorials/nextjs-analytics).
    
</CalloutBox>
```

```js
// instrumentation-client.js
import posthog from 'posthog-js'

posthog.init(process.env.NEXT_PUBLIC_POSTHOG_KEY, {
  api_host: process.env.NEXT_PUBLIC_POSTHOG_HOST,
  defaults: '2025-05-24',
});
```

After you set this up and restart your app, PostHog starts autocapturing events and pageviews. If you have enabled session replays in [your project settings](https://us.posthog.com/settings/project-replay), these will also be captured. 

<ProductScreenshot
  imageLight="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_07_15_at_13_11_31_2x_8687595a94.png"
  imageDark="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_07_15_at_13_12_00_2x_27d9fe2fca.png"
  alt="Events in PostHog"
  classes="rounded"
/>

<Callout type="info" title="Setting up a reverse proxy ">

We highly recommend setting up a reverse proxy to ensure requests aren't blocked by tracking blockers. You can find instructions for doing this in our [reverse proxy guide](/docs/advanced/proxy).

> </Callout>

Author: ```suggestion
</CalloutBox>
```

Simply, import PostHog into your `page.js` and call `posthog.capture()` when the button is clicked like this:

```js
// app/page.js
'use client'
import styles from './page.module.css'
import Link from 'next/link'
import posthog from 'posthog-js'

export default function Home() {

  const captureButtonClick = () => {
    posthog.capture('button_clicked', {
      cool: true
    })
  }

  return (
    <main className={styles.main}>
      <h1>Home</h1>
      <Link href="/about">Go to About</Link>
      <button onClick={captureButtonClick}>Click me</button>
    </main>
  )
}
```

## Setting up feature flags

You can use feature flags to control what users see in your app. For example, you can show a new feature to a subset of users and then roll it out to all users over time.

To set one up, go to the [feature flags page](https://us.posthog.com/feature_flags) in PostHog and create a new feature flag. Give it a key (I chose `main-cta`), make sure to roll it out to 100% of users, and click **Save**. 

<ProductScreenshot

>   imageLight="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_07_15_at_13_44_15_2x_485233b39a.png"
>   imageDark="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_07_15_at_13_44_38_2x_1010000000.png"

Author: ```suggestion
  imageLight="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_07_15_at_13_44_35_2x_8e7e69a22a.png"
  imageDark="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_07_15_at_13_44_15_2x_485233b39a.png"
```

``` file=.env.local
NEXT_PUBLIC_POSTHOG_KEY=<ph_project_api_key>
NEXT_PUBLIC_POSTHOG_HOST=<ph_client_api_host>
```

We can now initialize PostHog. Next.js 15.3 added support for `instrumentation-client.js|ts` which we can use to initialize PostHog. Create a `instrumentation-client.js` file in the base directory and add the following code:

```js
// instrumentation-client.js
import posthog from 'posthog-js'

posthog.init(process.env.NEXT_PUBLIC_POSTHOG_KEY, {
  api_host: process.env.NEXT_PUBLIC_POSTHOG_HOST,
  defaults: '2025-05-24',
});
```

After you set this up and restart your app, PostHog starts autocapturing events and pageviews. If you have enabled session replays in [your project settings](https://us.posthog.com/settings/project-replay), these will also be captured. 

<ProductScreenshot
  imageLight="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_07_15_at_13_11_31_2x_8687595a94.png"
  imageDark="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_07_15_at_13_12_00_2x_27d9fe2fca.png"
  alt="Events in PostHog"
  classes="rounded"
/>

> import { CalloutBox } from 'components/Docs/CalloutBox'

Author: ```suggestion
```

export default function PostHogClient() {
  const posthogClient = new PostHog(process.env.NEXT_PUBLIC_POSTHOG_KEY, {
    host: process.env.NEXT_PUBLIC_POSTHOG_HOST,
    flushAt: 1,

>     flushInterval: 0

Author: Because it can be used on short running processes like Vercel Functions
